src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs

```diff
@@ -10441,14 +10441,7 @@ internal Task TdsExecuteSQLBatch(string text, int timeout, SqlNotificationReques
                 }
                 catch (Exception e)
                 {
-                    // UNDONE - should not be catching all exceptions!!!
-                    if (!ADP.IsCatchableExceptionType(e))
```

Comment:
Is this the right thing to do? The code was earlier checking if the exception is not catchable and then throw it instead of doing a FailureCleanup. I suspect this may cause issues in some other scenarios down the line. 
Can't this be made more specific to the use case you are trying to tackle?

Reply:
Are you looking to handle NullReferenceException here ?

Reply:
If we are trying to handle NullReferenceException, then we should be looking at fixing the root cause of the NRE.

Reply:
The exception caught here comes from delegate implemented by user. In this case, the exception in first "bad" case occurs here: https://github.com/dotnet/SqlClient/pull/1068/files#diff-96a138d0ea44725f6c47597239572099a62c531a08ab225ed51e961ff2362505R104

Which means, user can introduce any exception and no matter what we handle we cannot cover all exceptions.

Furthermore, this change is in "TdsExecuteRPC", and if any exception occurs in here the command execution has failed, the StateObj should ideally be cleaned up for next use, since we cache StateObj in the driver and next command will pick it up.

Reply:
That sounds reasonable to me. Can you put a comment explaining that in the catch block so that no-one is tempted to revert the change? Just explain that user code can bubble exceptions to that point and we can't protect users against themselves in that case or something.

Reply:
@cheenamalhotra it looks like the user is passing in invalid values, causing the exception. I am trying to understand how the user can introduce *any* exception. 
Can you explain the context a little bit more around why we received an NRE in the first place when user added the null value? Is this really a user error since null can be passed as a valid input for data. Is the NRE due to a response from the server, which we failed to classify or is it because we got a null value which was de-referenced before sending it to the server?

Reply:
BTW the intention of my questions is to understand the scenario better. My hunch is that your fix may be the only way to go, however if I have to port it to .Net framework (system.data.dll), I am trying to understand the scenario correctly and so that I can defend this. If in this discussion, we realize that a more localized fix can be made for this use case, then that would be ideal. 
For .Net framework, most likely, I will be adding an app context switch where this fix is opt-in since the fix touches something fundamental. I believe, with your fix, for *any* error, we are trying to discard the connection and I wonder if for some large scale applications that would cause pool performance issues.
If we receive more cases for this issue, then I could, in future, consider turning on this fix in .Net framework by default. 
These decisions would be guided by your assessment on whether this is an actual user error or not. So far my understanding is that passing in a "null" may not be a user error.

Reply:
The action delegate to fetch 'id' and 'value' from 'x' (as pasted above) is user defined, where NRE occurs.

If you debug this test, you'll notice it's surface area.

Reply:
The part I am struggling to understand is that if there is a NRE in the case above, then the call stack shouldn't come from the driver, which is what the CSS had reported. 

```
BadFunc Exception is Object reference not set to an instance of an object.
   at System.Data.SqlClient.TdsParser.TdsExecuteRPC(SqlCommand cmd, _SqlRPC[] rpcArray, Int32 timeout, Boolean inSchema, SqlNotificationRequest notificationRequest, TdsParserStateObject stateObj, Boolean isCommandProc, Boolean sync, TaskCompletionSource`1 completion, Int32 startRpc, Int32 startParam)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, Boolean inRetry, SqlDataReader ds, Boolean describeParameterEncryptionRequest)
```

Reply:
I had a chat with Cheena, and she provided the explanation for the call stack.